clang/docs/UsersManual.rst

```diff
@@ -1202,6 +1202,30 @@ Clang also allows you to push and pop the current warning state. This is
 particularly useful when writing a header file that will be compiled by
 other people, because you don't know what warning flags they build with.
 
+Note that the following diagnostic groups, which are ones based on analyzing
```

Comment:
Ooof... this is a shocking/bad behavior.  I wish we could do better, but I of course know we probably couldn't?  UNLESS... we set the 'state' of the warning on the function during evaluation of the pragma, then checked that flag AS WELL at the end of TU to enable them?

Not sure if we want to do that in THIS patch, but an issue report with this idea isn't perhaps a terribad idea.

Reply:
I may be misunderstanding your suggestion, but I think that would end up in the same place. We'd see the push, set the state on the function, see the pop, clear the state on the function (which we'd need to do for correctness since these can be placed arbitrarily, including nesting), then get to the `}` and have missed the fact that the pragma was enabled for the lines we wanted to diagnose.

Really, the only way around this behavior that I think works is to not skip analysis-based warnings if the diagnostic is disabled at the end of the function. But that has some significant compile-time overhead because these are CFG-based analyses.

Reply:
Ah, no my suggestion is "see the push(actually the enable), set the state on the function".  "See the pop, do nothing other than change the per-line state".  The state is "was this EVER enabled in this function", not "IS this enabled now".

Then at end of function, you check "was ever enabled" kinda thing, then do the analysis.  THEN only emit the warnings that were actually enabled based on effective-line.

Reply:
Ah thanks for the explanation, I'll play around with that idea real quick

Reply:
Alternatively, maybe we can detect and diagnose `#pragma clang diagnostic` inside a function being analyzed and warn that it's ignored